### PR TITLE
Kevincai/back cart

### PR DESCRIPTION
**Pull request by kevinjcai**
[//]: # "These comments are meant for your reference. They are invisible and don't need to be deleted!"

### Description
[//]: # "Required - Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."

Back end cart

### Screenshots
[//]: # "Required for frontend changes, otherwise optional but strongly recommended. Add screenshots of expected behavior - GIFs if you're feeling fancy!"

## How to review
[//]: # "Required - Describe the order in which to review files and what to expect when testing locally. Is there anything specifically you want feedback on? Should this be reviewed commit by commit, or all at once? What are some user flows to test? What are some edge cases to look out for?"

## Next steps
[//]: # "Optional - What's NOT in this PR, doesn't work yet, and/or still needs to be done. Note any temporary fixes in this PR that should be cleaned up later."

### Online sources
[//]: # "Optional - copy links to any tutorials or documentation that was useful to you when working on this PR"

### Related PRs
[//]: # "Optional - related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

[//]: # "This tags the project leader as a default. Feel free to change, or add on anyone who you should be in on the conversation."
🛍 CC: @EthanAuyeung

### CODE REVIEW

**Comment by kevinjcai** [comment]
Remove the page.tsx before mergin 
Its a test command

File: src/api/supabase/queries/user_queries.ts [hunk 1]
```diff
@@ -16,87 +16,166 @@ const supabaseApiKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY;
...
+
+  // Use the updateCart function to update the cart in the database
+  await updateCart(userId, currentCart);
 }
+
+export async function decrementCartItemByOne(userId: string, itemId: string) {
+  return decrementCartItem(userId, itemId, 1);
+}
+
+// export async function addUserAddress(
```
Comment (surajr10) [inline src/api/supabase/queries/user_queries.ts]: remove all commented code

File: src/app/page.tsx [hunk 2]
```diff
@@ -23,20 +21,17 @@ import {
...
-  // testGetOrderById();
-  // testToggleOrderProgress();
-  // testFetchProducts();
-  // testFetchProductByName();
-  // testFetchPickupData();
-  // testFetchPickupTimesByUUID();
-  // testUpdateAllOrdersProgressToTrue();
+export const revalidate = 10;
+
+export default async function Checkout() {
```
Comment (surajr10) [inline src/app/page.tsx]: remove the async

Comment (surajr10) [inline src/app/page.tsx]: this is the root page of the project, i might rename the component from Checkout to something like Home

File: src/app/page.tsx [hunk 3]
```diff
@@ -23,20 +21,17 @@ import {
...
-  // testFetchOrderByUUID();
-  // testFetchOrders();
-  // testGetOrderById();
-  // testToggleOrderProgress();
-  // testFetchProducts();
-  // testFetchProductByName();
-  // testFetchPickupData();
-  // testFetchPickupTimesByUUID();
-  // testUpdateAllOrdersProgressToTrue();
+export const revalidate = 10;
```
Comment (surajr10) [inline src/app/page.tsx]: can remove

File: src/schema/schema.ts [hunk 4]
```diff
@@ -5,15 +5,16 @@ export type User = {
   first_name: string;
   last_name: string;
   pet_information: string;
-  delivery_allowed: boolean;
-  created_at: Date; // timestamp of when record was created
-  cart_id: string; // UUID
+  order_option: boolean;
+  creaed_at: string; // timestamp with time zone not null default now();
```
Comment (surajr10) [inline src/schema/schema.ts]: change to created_at

File: src/api/supabase/queries/user_queries.ts [hunk 5]
```diff
@@ -16,87 +16,166 @@ const supabaseApiKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY;
...
+      delete currentCart[itemId];
+    }
+  }
+
+  // Use the updateCart function to update the cart in the database
+  await updateCart(userId, currentCart);
 }
+
+export async function decrementCartItemByOne(userId: string, itemId: string) {
+  return decrementCartItem(userId, itemId, 1);
```
Comment (surajr10) [inline src/api/supabase/queries/user_queries.ts]: think you can just call `decrementCartItem` w/ out the return. additionally, i would await `decrementCartItem` b/c it's an `async` function

File: src/api/supabase/queries/user_queries.ts [hunk 6]
```diff
@@ -16,87 +16,166 @@ const supabaseApiKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY;
...
+}
+
+export async function decrementCartItem(
+  userId: string,
+  itemId: string,
+  n: number,
+) {
+  // First, fetch the current cart for the user
+  const { data, error }: { data: User[] | null; error: PostgrestError | null } =
+    await supabase.from('profiles').select('*').eq('user_id', userId);
```
Comment (surajr10) [inline src/api/supabase/queries/user_queries.ts]: can append `.single()` to end of your query so that you don't need to do the `data.length === 0` check , you can set `currentCart` to just `data.cart`, and you can change the type of data to just be `User` instead of `User[]`

File: src/api/supabase/queries/user_queries.ts [hunk 7]
```diff
@@ -16,87 +16,166 @@ const supabaseApiKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY;
...
+  // Use the updateCart function to update the cart in the database
+  await updateCart(userId, currentCart);
+}
 
-    return { data, error: null, status: 200, statusText: 'OK', count: 1 };
-  } catch (error) {
-    console.error('Error:', error);
-    throw error;
+export async function incrementCartItemByOne(userId: string, itemId: string) {
+  return incrementCartItem(userId, itemId, 1);
```
Comment (surajr10) [inline src/api/supabase/queries/user_queries.ts]: same comment that i left for `decrementCartItemByOne` function

File: src/api/supabase/queries/user_queries.ts [hunk 8]
```diff
@@ -16,87 +16,166 @@ const supabaseApiKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY;
...
+  } else {
+    return data;
+  }
+}
+
+export async function fetchUserCart(
+  userId: string,
+): Promise<Record<string, number>> {
+  const { data, error }: { data: User | null; error: PostgrestError | null } =
+    await supabase.from('profiles').select('*').eq('user_id', userId).single();
```
Comment (surajr10) [inline src/api/supabase/queries/user_queries.ts]: can replace this query w/  call to `fetchUserByUUID` function

File: src/api/supabase/queries/user_queries.ts [hunk 9]
```diff
@@ -16,87 +16,166 @@ const supabaseApiKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY;
...
+}
+
+export async function incrementCartItem(
+  userId: string,
+  itemId: string,
+  n: number,
+) {
+  // First, fetch the current cart for the user
+  const { data, error }: { data: User | null; error: PostgrestError | null } =
+    await supabase.from('profiles').select('*').eq('user_id', userId).single();
```
Comment (surajr10) [inline src/api/supabase/queries/user_queries.ts]: can replace this query w/ call to `fetchUserByUUID` function

File: src/api/supabase/queries/user_queries.ts [hunk 10]
```diff
@@ -16,87 +16,166 @@ const supabaseApiKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY;
...
+      `An error occurred when trying to fetch the cart: ${error.message}`,
+    );
+  } else if (!data) {
+    throw new Error('No user found with the specified user_id.');
+  }
+
+  const currentCart = data.cart;
+
+  // Increment the item's quantity by n or set it to n if not present
+  currentCart[itemId] = (currentCart[itemId] || 0) + n;
```
Comment (surajr10) [inline src/api/supabase/queries/user_queries.ts:40]: this is cleann

File: src/api/supabase/queries/user_queries.ts [hunk 11]
```diff
@@ -16,87 +16,166 @@ const supabaseApiKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY;
...
-    const { data: users, error } = await supabase
-      .from('profiles')
-      .select('*')
-      .single();
-
-    if (error) {
-      console.error('Error fetching data:', error);
-      return { data: [], error };
-    }
+export async function fetchUserData() {
```
Comment (surajr10) [inline src/api/supabase/queries/user_queries.ts]: make sure to type return values of all your functions

Comment (surajr10) [inline src/api/supabase/queries/user_queries.ts]: is there a use case for fetching all users? if not you can probably delete this function. if so, nit but i'd suggest renaming to fetchAllUsers

File: src/api/supabase/queries/user_queries.ts [hunk 12]
```diff
@@ -16,87 +16,166 @@ const supabaseApiKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY;
...
+  const { data, error }: { data: User[] | null; error: PostgrestError | null } =
+    await supabase.from('profiles').select('*');
 
-    return { data: users } as PostgrestSingleResponse<User[]>;
-  } catch (error) {
-    console.error('Error:', error);
-    throw error;
+  if (error) {
+    throw new Error(`An error occured when trying to read profiles: ${error}`);
```
Comment (surajr10) [inline src/api/supabase/queries/user_queries.ts]: [nit] standardize the errors for all functions. i think some are using ${error} and some are using ${error.message}. i think either is fine honestly

File: src/api/supabase/queries/user_queries.ts [hunk 13]
```diff
@@ -16,87 +16,166 @@ const supabaseApiKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY;
...
+    .eq('user_id', uuid)
+    .single();
+  if (error) {
+    throw new Error(`An error occured when trying to read profiles: ${error}`);
+  } else {
+    return data;
+  }
+}
+
+export async function fetchUserCart(
```
Comment (surajr10) [inline src/api/supabase/queries/user_queries.ts]: i'd say for all queries that are rn taking the `userId` as a parameter, you can remove it. i don't think yall are using an auth/profile context, so i'd say when you're calling these queries from your components, it's probably easier to just omit getting the `userId` there and passing it to the queries. rather i'd say it'd be easier to just use the `supabase.auth.getUser()` function (which returns current logged in user which makes sense b/c all of these queries should only be available to person who's logged in at the moment) inside these queries and you can get the `userId` off of the object returned here.

**Review by surajr10 — COMMENTED** [review]
great work on this! lots of clean queries! lmk if you got any questions about anything i wrote. i'd prob also just do a run through to see if all the queries are working again after going through any changes. also make sure to fill out the PR description before merging!

**Comment by EthanAuyeung** [comment]
Refactored
